Sema: Desugar non-generic typealiases with an unbound generic underlying type

slavapestov opened the pull request:
We let you write this:

```
  typealias Foo = Array
```

This had the same exact interpretation, but a different representation than,
the generic equivalent:

```
  typealias Foo<Element> = Array<Element>
```

Now, we desugar the former to the latter by lazily synthesizing a
GenericParamList as needed for a typealias of this kind, followed by
implicitly applying the generic arguments in UnderlyingTypeRequest.

slavapestov committed:
Sema: Add skipRequirementsCheck parameter to applyUnboundGenericArguments()

slavapestov committed:
Sema: Desugar non-generic typealiases with an unbound generic underlying type

We let you write this:

  typealias Foo = Array

This had the same exact interpretation, but a different representation than,
the generic equivalent:

  typealias Foo<Element> = Array<Element>

Now, we desugar the former to the latter by lazily synthesizing a
GenericParamList as needed for a typealias of this kind, followed by
implicitly applying the generic arguments in UnderlyingTypeRequest.

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

AnthonyLatsis commented on test/Generics/function_decls.swift:
```diff
@@ -44,7 +44,7 @@ func f14<T : P>(x: T) -> T.C<Int> {}
 func f15<T : P>(x: T) -> T.D<Int> {}
 
-func f16<T : P>(x: T) -> T.D {} // expected-error {{reference to generic type 'T.D' (aka 'G') requires arguments in <...>}}
+func f16<T : P>(x: T) -> T.D {} // expected-error {{reference to generic type 'T.D' requires arguments in <...>}}
```
Perhaps these aliases still deserve an 'aka' so that the source of generics is immediately clear to the user.

slavapestov replied:
Yeah, I’ll look into that. The reason we don’t do that now is because generic typealiases use an UnboundGenericType here instead of TypeAliasType, which is also weird.

AnthonyLatsis replied:
Oh, I see. On the other hand, it would be dangerous to allow something like `typealias Foo<T> = T` to be casually desugared before being fully bound.

AnthonyLatsis commented on test/decl/typealias/unbound.swift:
```diff
@@ -0,0 +1,49 @@
+  // expected-error@-3 {{reference to generic type 'Array' requires arguments in <...>}}
+
+  typealias WeirdestType = Array where T : Equatable
+  // expected-error@-1 {{reference to generic type 'Array' requires arguments in <...>}}
```
This is a source-breaking limitation; is there a problem? I wish we could ban this kind of implicit genericity sugar altogether, though.

slavapestov replied:
It appears that we used to just ignore a `where` clause on this kind of typealias in the past. So if there's a source break, the code was already invalid because the `where` clause wasn't checked.

slavapestov committed:
Sema: More elaborate hack for rdar://problem/53164203

The old hack was not sufficient one we begin desugaring typealiases
with unbound generic underlying types to generic typealiases.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

slavapestov committed:
Sema: Fix extensions of unbound generic typealiases

slavapestov commented:
@swift-ci Please smoke test
